airbyte-commons-worker/src/main/java/io/airbyte/workers/RecordSchemaValidator.java

```diff
@@ -52,31 +62,55 @@ public RecordSchemaValidator(final FeatureFlagClient featureFlagClient,
       ((ObjectNode) schema).put("$schema", "http://json-schema.org/draft-07/schema#");
       validator.initializeSchemaValidator(stream.toString(), schema);
     }
-
   }
 
   /**
    * Takes an AirbyteRecordMessage and uses the JsonSchemaValidator to validate that its data conforms
    * to the stream's schema If it does not, this method throws a RecordSchemaValidationException
    *
-   * @param message
    * @throws RecordSchemaValidationException
    */
-  public void validateSchema(final AirbyteRecordMessage message, final AirbyteStreamNameNamespacePair messageStream)
-      throws RecordSchemaValidationException {
-
-    final JsonNode messageData = message.getData();
-    final JsonNode matchingSchema = streams.get(messageStream);
+  public void validateSchema(
+                             final AirbyteRecordMessage message,
+                             final AirbyteStreamNameNamespacePair messageStream,
+                             final ConcurrentHashMap<AirbyteStreamNameNamespacePair, ImmutablePair<Set<String>, Integer>> validationErrors) {
+    if (backgroundValidation) {
+      validationExecutor.execute(() -> {
+        try {
+          doValidateSchema(message, messageStream);
+        } catch (final RecordSchemaValidationException e) {
+          handleException(e, messageStream, validationErrors);
+        }
+      });
+    } else {
+      try {
+        doValidateSchema(message, messageStream);
+      } catch (final RecordSchemaValidationException e) {
+        handleException(e, messageStream, validationErrors);
+      }
+    }
+  }
 
-    if (workspaceId != null) {
-      if (featureFlagClient.enabled(PerfBackgroundJsonValidation.INSTANCE, new Workspace(workspaceId))) {
-        log.debug("feature flag enabled for workspace {}", workspaceId);
+  private void handleException(final RecordSchemaValidationException e,
+                               final AirbyteStreamNameNamespacePair messageStream,
+                               final ConcurrentHashMap<AirbyteStreamNameNamespacePair, ImmutablePair<Set<String>, Integer>> validationErrors) {
+    validationErrors.compute(messageStream, (k, v) -> {
+      if (v == null) {
+        return new ImmutablePair<>(e.errorMessages, 1);
       } else {
-        log.debug("feature flag disabled for workspace {}", workspaceId);
+        final var updatedErrorMessages = Stream.concat(v.getLeft().stream(), e.errorMessages.stream()).collect(Collectors.toSet());
+        final var updatedCount = v.getRight() + 1;
+        return new ImmutablePair<>(updatedErrorMessages, updatedCount);
       }
-    } else {
-      log.debug("workspace id is null");
-    }
+    });
+  }
+
+  /**
+   * @throws RecordSchemaValidationException
+   */
+  private void doValidateSchema(final AirbyteRecordMessage message, final AirbyteStreamNameNamespacePair messageStream) {
```

Comment:
nit: instead of `messageStream` -> `airbyteStream`?

`messageStream` keeps on making me think this is a stream of messages.